src/NServiceBus.Transport.RabbitMQ/Connection/ChannelProvider.cs

```diff
@@ -21,34 +21,47 @@ public ChannelProvider(ConnectionFactory connectionFactory, TimeSpan retryDelay,
 
         public void CreateConnection()
         {
-            connection = connectionFactory.CreatePublishConnection();
+            connection = connectionFactory.CreatePublishConnection(); // Can take over 5 seconds
             connection.ConnectionShutdown += Connection_ConnectionShutdown;
         }
 
         void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
         {
-            if (e.Initiator != ShutdownInitiator.Application)
+            if (e.Initiator == ShutdownInitiator.Application)
             {
-                var connection = (IConnection)sender;
-
-                // Task.Run() so the call returns immediately instead of waiting for the first await or return down the call stack
-                _ = Task.Run(() => ReconnectSwallowingExceptions(connection.ClientProvidedName), CancellationToken.None);
+                return;
             }
+
+            var connectionThatWasShutdown = (IConnection)sender;
```

Comment:
From just reading the code, I'm not sure if it is possible for this event handler to run for any other connection apart from the one referenced by the `connection` field?

In any case, if `sender` and `connection` are the same, should we set `connection` to null after disposing it?  This would enable us to avoid trying to use a disposed connection in `GetPublishChannel()`, which looks like it is currently possible.

Reply:
@abparticular I think we had some discussion about this in https://github.com/Particular/NServiceBus.RabbitMQ/pull/1270#discussion_r1239042012 but then eventually asynchronously concluded it is not strictly required, so we removed the explicit null assignment because the code today assumes when things are disposed nothing will acquire channels anymore.

Since the existing code already operated on the event arguments, I wanted to keep that in place to not make any other assumptions.

The downside of setting it to null in the case someone would acquire a publish channel is that you'd get NullRefs while today you would at least get some for of stack trace that shows a publish channel trying to access a disposed connection which might be more revealing should we ever run into such an edge case. 

What do you think?

Reply:
> the code today assumes when things are disposed nothing will acquire channels anymore

I don't understand how that assumption is valid.  Won't the message dispatcher potentially continue trying to get channels while the reconnection loop is running?

> The downside of setting it to null in the case someone would acquire a publish channel is that you'd get NullRefs while today you would at least get some for of stack trace that shows a publish channel trying to access a disposed connection which might be more revealing should we ever run into such an edge case.
>
> What do you think?

Was it considered to make GetPublishChannel() async and only return a channel after reconnection is successful (ie. when connection is not null)?